src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,678 @@
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use.
+     *     See the {@code KDF} section in the <a href=
+     *     "{@docRoot}/../specs/security/standard-names.html#kdf-algorithms">
+     *     Java Security Standard Algorithm Names Specification</a>
+     *     for information about standard KDF algorithm names.
+     * @param kdfParameters
+     *     the {@code KDFParameters} used to configure the derivation
+     *     algorithm or {@code null} if no parameters are provided
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if no {@code Provider} supports a {@code KDF} implementation for the
+     *     specified algorithm
+     * @throws InvalidAlgorithmParameterException
+     *     if no {@code Provider} supports a {@code KDFSpi} implementation for
+     *     the specified algorithm and parameters
+     * @throws NullPointerException
+     *     if the {@code algorithm} is {@code null}
+     */
+    public static KDF getInstance(String algorithm,
+                                  KDFParameters kdfParameters)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Objects.requireNonNull(algorithm, "algorithm must not be null");
+        // make sure there is at least one service from a signed provider
+        Iterator<Service> t = GetInstance.getServices("KDF", algorithm);
+        NoSuchAlgorithmException lastException = null;
+        while (t.hasNext()) {
+            Service s = t.next();
+            if (!JceSecurity.canUseProvider(s.getProvider())) {
+                continue;
+            }
+            try {
+                Object obj = s.newInstance(kdfParameters);
+                if (!(obj instanceof KDFSpi spiObj)) {
+                    lastException = new NoSuchAlgorithmException(
+                        new InvalidAlgorithmParameterException(
+                            "No provider can be found that supports the "
+                            + "specified parameters"));
+                    if (!skipDebug && pdebug != null) {
+                        pdebug.println(
+                            "obj was not an instance of KDFSpi (should not "
+                            + "happen)");
+                    }
+                    // continue to next iteration
+                } else if (t.hasNext()) {
+                    return new KDF(new Delegate(spiObj, s.getProvider()), t,
+                                   algorithm, kdfParameters);
+                } else { // no other choices, lock down provider
+                    return new KDF(new Delegate(spiObj, s.getProvider()),
+                                   algorithm, kdfParameters);
+                }
+            } catch (NoSuchAlgorithmException e) {
+                lastException =
+                    new NoSuchAlgorithmException(
+                        new InvalidAlgorithmParameterException(
```

Comment:
Slight tweak in message: s/specified parameters/specified algorithm and parameters/